openlmis-stock/src/main/java/org/smartregister/stock/openlmis/intent/service/ReasonSyncIntentService.java

```diff
@@ -31,13 +37,20 @@ public int onStartCommand(Intent intent, int flags, int startId) {
 
     @Override
     protected void onHandleIntent(Intent workIntent) {
+        String facilityTypeUuid = OpenLMISLibrary.getInstance().getFacilityTypeUuid();;
         if (NetworkUtils.isNetworkAvailable(context)) {
-            pullFromServer();
+            // assumes eventual consistency where all programs are synced and reasons for all programs can be fetched
+            for (Program program : OpenLMISLibrary.getInstance().getProgramRepository().findAllPrograms()) {
+                if (facilityTypeUuid != null) {
+                    pullFromServer(REASON_SYNC_URL + "?" + FACILITY_TYPE_UUID + "=" + facilityTypeUuid + "&" + PROGRAM_ID + "=" + program.getId());
+                }
+            }
+            pullFromServer(REASON_SYNC_URL + "?");
```

Comment:
This should be inside the for loop as an else. Better yet would be invoke server once with a list of programIds